projects/clr/hipamd/include/hip/amd_detail/amd_hip_fp8.h

```diff
@@ -949,6 +950,122 @@ __FP8_HOST_STATIC__ __hip_fp8x2_storage_t __hip_cvt_halfraw2_to_fp8x2(
   return __hip_cvt_float2_to_fp8x2(__half22float2(__half2(x)), sat, interp);
 }
 
+namespace hip_detail {
+
+constexpr __hip_fp8_storage_t e8m0_NaN = 0xFFU;
+constexpr __hip_internal::uint16_t bf16_NaN = 0x7FFFU;
+
+constexpr __hip_internal::uint16_t bf16_sig_mask = 0x007FU;
+constexpr __hip_internal::uint32_t float_sig_mask = 0x007FFFFFU;
+constexpr __hip_internal::uint64_t double_sig_mask = 0x000FFFFFFFFFFFFFU;
+
+constexpr __hip_internal::uint16_t bf16_max_exp = 0x7F80U;
+constexpr __hip_internal::uint32_t float_max_exp = 0x7F800000U;
+constexpr __hip_internal::uint64_t double_max_exp = 0x7FF0000000000000U;
+
+constexpr __hip_internal::uint16_t bf16_sign_mask = 0x8000U;
+constexpr __hip_internal::uint32_t float_sign_mask = 0x80000000U;
+constexpr __hip_internal::uint64_t double_sign_mask = 0x8000000000000000U;
+
+constexpr __hip_internal::uint16_t bf16_half_sig_bit = 0x0040U;
+constexpr __hip_internal::uint32_t float_half_sig_bit = 0x00400000U;
+constexpr __hip_internal::uint64_t double_half_sig_bit = 0x0008000000000000U;
+
+}  // namespace hip_detail
+
+__FP8_HOST_DEVICE_STATIC__ __hip_fp8_storage_t __hip_cvt_double_to_e8m0(
+    const double val, const __hip_saturation_t saturate, const enum hipRoundMode rounding) {
+  union {
+    double as_double;
+    __hip_internal::uint64_t as_int;
+  } u{val};
+
+  // Shifts out mantissa bits from double dtype
+  unsigned short double_exp =
+      static_cast<unsigned short>((~hip_detail::double_sign_mask & u.as_int) >> 52);
+  __hip_fp8_storage_t e8m0;
+  if (double_exp == 0x0U) {
+    e8m0 = 0x0U;
+  } else if ((double_exp - 0x0380U) > 0x00FF) {  // if double is NaN/Inf/or too large
+    e8m0 = hip_detail::e8m0_NaN;
+  } else {
+    e8m0 =
+        double_exp - 0x0380U;  // shift due to bias difference between double and single precision
+  }
+
+  // If there is a mantissa and the exp wont overflow round up
+  if ((rounding == hipRoundPosInf) && (u.as_int & hip_detail::double_sig_mask) &&
```

Comment:
I'm not clear on what rounding modes are valid here.  Are rounding to -Inf or toward/away from zero not valid?